### ISSUE TITLE
Expose clientManifest in pageContext during production runtime

### ISSUE THREAD

**Issue by Csszabi98** [brillout/vite-plugin-ssr#732]
### Description

As a consumer of `vite-plugin-ssr` I would like to be able to access the "clientManifest" in an easy way (vite-plugin-ssr already loads it for internal use), to allow preloading my dynamic assets. The library already provides static dependency analyses but this is insufficient for more dynamic content-driven use cases.

**This would be useful when:**
-  you can get a reference to used assets during a render (as described in https://vitejs.dev/guide/ssr.html#generating-preload-directives)
-  you will know ahead of time which components will be rendered on a dynamic page. (Analyzing preliminary API response)

I would even be in favor of passing the references to vite plugin ssr of these preloadable assets instead of manually generating all the preload directives and injecting them in the html. That way we could easily avoid asset preload duplication without the use of `injectFilter`.

Regardless, clientManifest would be vital for such a feature to be able to infer the manifestKeys using the plain component names.

**Comment by brillout** [comment]
I agree in principle, although so far I think the DX can be better instead of just providing the content of `manifest.json`.

> I would like to be able to access the "clientManifest" [...] to allow preloading my dynamic assets. [...] The library already provides static dependency analyses but this is insufficient for more dynamic content-driven use cases.

When and what kind of assets do you want to preload? The more I know about your use case the better.


> I would even be in favor of passing the references to vite plugin ssr of these preloadable assets instead of [...] `injectFilter`.

What issues are you having with `injectFilter`?

Note that, for various reasons, there can be only two positions: at the begin (before the HTML of `pageContext.Page`) and at the end (after the HTML of `pageContext.Page`). That's why https://github.com/brillout/vite-plugin-ssr/issues/695 has exactly two placeholders.

**Comment by Csszabi98** [comment]
My use case is the following:

- Only parts of my web page are rendered on the server (above the fold mainly), the rest is loaded on the client side. This is due to the content being highly dynamic and some system level limitations which disallowing the page regeneration when some parts of the content (shared central content) change.
- Due to this I only use a single entry (page file) to render the dynamic content depending on the given page's structure. This means that the manifest json cannot really be statically analyzed as the app is littered with dynamic imports.
- This leads to me using vite's built in module tracking which works wonderfully, although, I miss most of my components being preloaded due to most of them rendering on the client only.
- To work around this, I analyze the full page structure on the server side during the cached render, and use the basic data of x, y component will be rendered to generate asset preload directives for my components as they will be rendered always.
- As for what I want to preload it varies on what the component relies upon. It can be images, css, javascript.
- For the injectFilter my only problem is the warning which cannot be turned off when I don't inject certain assets due to me already including them directly on my own (this avoids double preload directives).

**Comment by Csszabi98** [comment]
For DX improvements, it would be nice to pass some sort of an object to the plugin instead of having to generate all the preload directives manually. How I would imagine this:

- I resolve the manifestKey which has to be preloaded
- I pass it to the plugin
- The plugin does the rest

**Comment by brillout** [comment]
Seems like a set of requirements that are very specific to your app.

In general, I consider implementing features that target only a single user only if that user is a sponsor. Maybe your employer would be up to sponsor VPS? (Or one time donation/contract but I prefer long-term sponsoring.)

**Comment by brillout** [comment]
Closing for lack of interest.
